### PR TITLE
Update AggregatorFactory to provide a method to indicate if it supports concurrent search

### LINKED ISSUE #9446: [Concurrent Segment Search] Dynamic cluster setting to disable concurrent segment search for a given aggregation type
As a part of https://github.com/opensearch-project/OpenSearch/issues/9316 it looks like it would be useful to have a way to disable concurrent segment search when certain aggregator types are used. This could be especially useful in the case of custom aggregator plugins that we do not have any knowledge of but due to how they are implemented either perform worse with or are completely unable to support concurrent aggregations.

A possible alternative would be to extend the `AggregatorFactory` with a `supportsConcurrentSearch()` method however this would require code level changes for any existing plugins while a dynamic cluster setting less intrusive to the plugins and should be able to be modified without any downtime or disruptions.

### PR DESCRIPTION
**Pull request by jed326**
Also refactored parent-join search ITs to support concurrent search parameterization and introduced a new test to cover concurrent search failures.

### Description
Some types of aggregations, especially those in custom search plugins, may be unable to utilize concurrent segment search due to their internal implementation. This setting provides a way to disable concurrent segment search when certain types of aggregations are used.

For the newly added tests, both of these tests will actually pass with concurrent search enabled due to problem 2 here https://github.com/opensearch-project/OpenSearch/issues/9316#issue-1850332933. However, once problem 2 is fixed these tests will catch the issue in problem 1. 

### Related Issues
Resolves #9446
Relates #9316

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jed326** [comment]
@reta @sohami opened a PR for #9446, could you please help review? Thanks!

**Comment by reta** [comment]
> @reta Thanks for the feedback! I briefly discussed that here [#9446 (comment)](https://github.com/opensearch-project/OpenSearch/issues/9446#issue-1857375837) (apologies the information is kind of spread out across a few issues). There are 2 main reasons why I think a dynamic cluster setting is preferable here:

I understand the tradeoffs of dynamic setting. Since at this moment we cannot guarantee that all aggregation out of the box are compatible with concurrent search, the safe option would be to enable them on per aggregation basis which will be done in backward compatible way (no issues here to have it in 2.x if implemented properly). The most beneficial thing to have it in API is awareness for plugin developers that there are additional considerations to take into account (and that would be done on explicit optin manner by plugin developers).

I assume the aggregation either works or does not work with concurrent search so flipping nature of dynamic setting is not helpful here. Plus, now we offload this burden to users, how they should know which aggregations are compatible with concurrent search and which are not? It is an option for sure but I think not the one we should implement.

**Comment by jed326** [comment]
> I assume the aggregation either works or does not work with concurrent search so flipping nature of dynamic setting is not helpful here. 

For the `parent` aggregation that kicked off this whole investigation, the "correctness" will actually depend on the cardinality of the data. If there is only 1 child document for every parent document, the current implementation will work just fine with concurrent search. In general it seems like the cardinality of the data can affect both the correctness and performance of aggregations, which is where the benefits of a dynamic setting come in.

> in backward compatible way (no issues here to have it in 2.x if implemented properly). The most beneficial thing to have it in API is awareness for plugin developers 

I'm splitting hairs a little bit here but I do think these two things are slightly at odds with each other. If we do this in a bwc way (default concurrent search support to false), then plugin developers won't necessarily be aware that there is a concurrent search support override they can add, especially for any existing plugins.

> Plus, now we offload this burden to users, how they should know which aggregations are compatible with concurrent search and which are not?

I think this is a good point, but at the same time ideally any custom search plugin should have functionality testing to determine if the results are still correct after enabling concurrent search.

**Comment by reta** [comment]
> For the parent aggregation that kicked off this whole investigation, the "correctness" will actually depend on the cardinality of the data

Please, we are not discussing the edge cases here (everything will work with empty index as an example), we should be mindful so correctness means - the aggregation return the correct results all the time.

> I'm splitting hairs a little bit here but I do think these two things are slightly at odds with each other. If we do this in a bwc way (default concurrent search support to false), then plugin developers won't necessarily be aware that there is a concurrent search support override they can add, especially for any existing plugins.

This is totally fine I think, remember that we would potentially require either additional work (or worst, complete rework), it is much safer (for 2.x) to follow the "least surprise". Also keep in mind, users would report bugs to plugin authors but we caused them. 

> I think this is a good point, but at the same time ideally any custom search plugin should have functionality testing to determine if the results are still correct after enabling concurrent search.

This is why we should make it API driven, not settings driven - plugin developer would not even notice there is a setting, and what should they do with it?

**Comment by sohami** [comment]
> I think this is a good point, but at the same time ideally any custom search plugin should have functionality testing to determine if the results are still correct after enabling concurrent search.

>> This is why we should make it API driven, not settings driven - plugin developer would not even notice there is a setting, and what should they do with it?

I agree that API level override in Aggregation will be better for figuring in backend core engine about which Aggregation doesn't work with concurrent segment search and decide which path to take based on it. By default, it will be `false` and once plugin developer tested the specific Aggregation with concurrent segment search enabled they can flip it to `true`. We may need to add some documentation around this to make it more visible to plugin developers and not miss on opportunity where existing plugin will work just fine.

For dynamic setting, the thought process was that mostly we are not seeing Aggregations having issues with concurrent segment search. This setting can help for cases where on a cluster user is using multiple different aggregations across request types and ended up hitting some bugs for a particular aggregation. So they will have the option to disable it for specific aggregation request types. On thinking more about this was wondering if we can instead make it a request level option for such use cases. Cluster level is global backend setting and request level can be an override for global setting for specific request types. Request level will be generic for any other operators and not specific to aggregations only.

To summarize: We can do the API level change for now and let the plugins specific implementation to enable once the have verified. We can explore the need for request level as a follow-up, meanwhile index/cluster level settings can be used to handle cases where some unknown issues are seen with concurrent search.

**Comment by reta** [comment]
Thanks @sohami 

> We may need to add some documentation around this to make it more visible to plugin developers and not miss on opportunity where existing plugin will work just fine.

Also, we may add log message to highlight that the concurrent path was not taken

> To summarize: We can do the API level change for now and let the plugins specific implementation to enable once the have verified. 

:+1:

**Comment by jed326** [comment]
@sohami @reta Thanks for the discussion!

I see now that coming at it from the plugin developer perspective, the dynamic cluster setting doesn't provide a way for the developer to indicate that their plugin does not support concurrent search. Moreover, the implementation I have here is using the "type" string of the aggregation, which is provided by the aggregation itself and can be re-used by different plugins (multiple aggregation with the same "type" just can't be installed at the same time). This means that there isn't really a good way to disable concurrent aggs for a given plugin by default.

Originally the thought was that there may be some undiscovered edge cases in some aggregations that do not work well with concurrent search and thus we should provide a way to dynamically disable concurrent search in those cases, however I agree that having this be a request level option is more user friendly. I'll create a separate issue for us to discuss the specifics of if/how we want a request level concurrent search parameter and I'll change this PR to have an API level check for concurrent search.

**Comment by jed326** [comment]
Pushed a new revision that adds a `supportsConcurrentSegmentSearch()` method to the `Aggregator` class which will be used in `AggregatorFactory::create` to update `searchContext.useConcurrentSearch`.

The default in `Aggregator` is `false` like we discussed but I have not implemented the override in all of the child classes yet since I wanted to check in and see if the general approach looks good first before doing so. Please let me know what you think @reta @sohami thanks!

**Comment by jed326** [comment]
Upon closer inspection, I don't think the latest commit will work because `Aggregator` objects are created from `AggregationProcessor::preProcess` which means this is too late in the request -- `isConcurrentSegmentSearchEnabled()` will have already been checked to create the `AggregationCollectorManager`. This means we will need to use implement `supportsConcurrentSegmentSearch` at the `AggregatorFactory` level and make sure we set the search context properly in `SearchService::parseSource`.

**Comment by jed326** [comment]
@reta @sohami this is ready for review now, let me know what you think. Thanks!

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java [hunk 1]
```diff
@@ -879,8 +881,20 @@ public Profilers getProfilers() {
...
+        // concurrentSearchSettingsEnabled is evaluated when searchContext is created
+        return concurrentSearchSettingsEnabled && requestShouldUseConcurrentSearch;
+    }
+
+    public void evaluateRequestShouldUseConcurrentSearch() {
+        requestShouldUseConcurrentSearch = true;
+        if (aggregations() != null && aggregations().factories() != null) {
+            for (AggregatorFactory factory : aggregations().factories().getFactories()) {
+                if (factory.traverseChildFactories() == false) {
+                    requestShouldUseConcurrentSearch = false;
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/DefaultSearchContext.java:895]: With the recommendation provided below we can change this to just do `return aggregations().factories(). allFactoriesSupportConcurrentSearch()`

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java [hunk 2]
```diff
@@ -879,8 +881,20 @@ public Profilers getProfilers() {
      * Returns concurrent segment search status for the search context
      */
     @Override
-    public boolean isConcurrentSegmentSearchEnabled() {
-        return useConcurrentSearch;
+    public boolean shouldUseConcurrentSearch() {
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/DefaultSearchContext.java:884]: lets add a java doc stating that should be used after request parsing ?

File: server/src/main/java/org/opensearch/search/SearchService.java [hunk 3]
```diff
@@ -1419,6 +1416,11 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             final CollapseContext collapseContext = source.collapse().build(queryShardContext);
             context.collapse(collapseContext);
         }
+        context.evaluateRequestShouldUseConcurrentSearch();
+        if (source.profile()) {
+            context.setProfilers(new Profilers(context.searcher(), context.shouldUseConcurrentSearch()));
+        }
+        logger.info("Using concurrent search for request: " + context.shouldUseConcurrentSearch());
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/SearchService.java:1423]: We can remove this log line.

File: server/src/main/java/org/opensearch/search/internal/SearchContext.java [hunk 4]
```diff
@@ -399,15 +399,15 @@ public final void assignRescoreDocIds(RescoreDocIds rescoreDocIds) {
     /**
      * Returns concurrent segment search status for the search context
      */
-    public boolean isConcurrentSegmentSearchEnabled() {
+    public boolean shouldUseConcurrentSearch() {
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/internal/SearchContext.java:402]: Can we make this abstract instead ?

File: server/src/main/java/org/opensearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java [hunk 5]
```diff
@@ -558,4 +558,8 @@ public String toString() {
         }
     }
 
+    @Override
+    protected boolean supportsConcurrentSegmentSearch() {
+        return true;
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java:563]: `return traverseChildFactories()`

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java [hunk 6]
```diff
@@ -114,4 +114,17 @@ public AggregatorFactory getParent() {
...
+        return false;
+    }
+
+    public boolean traverseChildFactories() {
+        for (AggregatorFactory aggregatorFactory : factories.getFactories()) {
+            if (aggregatorFactory.traverseChildFactories() == false) {
+                return false;
+            }
+        }
+        return supportsConcurrentSegmentSearch();
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java:128]: we can remove this and in the override of `supportsConcurrentSegmentSearch` each `AggregationFactory` that supports it will just call this method as suggested below.

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java [hunk 7]
```diff
@@ -114,4 +114,17 @@ public AggregatorFactory getParent() {
     public String getStatsSubtype() {
         return OTHER_SUBTYPE;
     }
+
+    protected boolean supportsConcurrentSegmentSearch() {
+        return false;
+    }
+
+    public boolean traverseChildFactories() {
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java:122]: `protected`

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactories.java [hunk 8]
```diff
@@ -257,6 +257,10 @@ private AggregatorFactories(AggregatorFactory[] factories) {
         this.factories = factories;
     }
 
+    public AggregatorFactory[] getFactories() {
+        return factories;
+    }
+
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/aggregations/AggregatorFactories.java:263]: instead of returning the `factories`, we can add a method `allFactoriesSupportConcurrentSearch` here as well which can iterate over all the `AggregationFactory`

```
public boolean allFactoriesSupportConcurrentSearch() {
for (AggregationFactory factory : factories) {
    if (!factory.supportsConcurrentSegmentSearch()) {
         return false;
    }
}
return true; 
}
```

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java (hunk 2 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:884]: I wonder if we should just explicitly enforce this actually. For example `requestShouldUseConcurrentSearch` could be a ` final java.lang.Boolean` then we initialize it only once in `evaluateRequestShouldUseConcurrentSearch` and if it's `null` we throw an exception in this method. What do you think?

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java (hunk 6 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java:128]: I like the suggestion above of `allFactoriesSupportConcurrentSearch` so we can move the iteration logic out of `DefaultSearchContext`, but I think the pattern of `supportsConcurrentSegmentSearch` being the recursive base case is a lot more intuitive -- when reading the code seeing `supportsConcurrentSegmentSearch()` returning `true` if it supports concurrent search is a lot more intuitive than returning `traverseChildFactories()`.

File: server/src/main/java/org/opensearch/search/internal/SearchContext.java (hunk 4 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/internal/SearchContext.java:402]: Going to leave this as concrete for now since SearchContext has a bunch of anonymous implementations in various test classes. We can revisit this later but for the purpose of this PR this should be fine.

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java (hunk 2 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:883]: Added an `assert` and javadoc in latest revision.

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java (hunk 6 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java]: I moved the `factories`logic out of `DefaultSearchContext` and now all of the iteration will happen in `AggregatorFactories` while `AggregatorFactory` will just perform the recursive call.

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java [hunk 9]
```diff
@@ -114,4 +114,12 @@ public AggregatorFactory getParent() {
     public String getStatsSubtype() {
         return OTHER_SUBTYPE;
     }
+
+    protected boolean supportsConcurrentSegmentSearch() {
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java:118]: nit: lets add java doc for the method something along the lines of:

```
Implementation should override this method and return true if the Aggregator created by the factory works with concurrent segment search execution model
```

**Review by sohami — COMMENTED** [review]
@jed326 Thanks for the changes. Overall LGTM other than a minor comment

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java [hunk 10]
```diff
@@ -876,11 +877,23 @@ public Profilers getProfilers() {
...
+        assert requestShouldUseConcurrentSearch != null : "requestShouldUseConcurrentSearch must be set";
+        return concurrentSearchSettingsEnabled && requestShouldUseConcurrentSearch;
+    }
+
+    /**
+     * Evaluate if parsed request supports concurrent segment search
+     */
+    public void evaluateRequestShouldUseConcurrentSearch() {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: I believe this method is not needed: we could use `aggregations(...)` setter from the search context to decide if concurrent search could be used. This is the only place those are set.

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java (hunk 9 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java:121]: Added!

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java (hunk 10 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: I'm having trouble finding the issue now but I believe @sohami is looking at disabling concurrent segment search for certain (non-agg) queries as well based on the request so I was coming at this from the perspective of introducing a way to determine if we should use concurrent segment search based on the request in general rather than just for aggs. @sohami please let me know if I'm wrong about this though and I'd be happy to make the change. Thanks!

Comment (reta) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: This is totally fine as well since we have `request` in the `DefaultSearchContext`  and could make this decision upfront, so there is no need to have dedicated method call to evaluate that

Comment (sohami) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: @reta This method will be a single place to evaluate for request based conditions where we want to disable fallback to non concurrent path for the request. There are cases due to which using `request` passed in constructor doesn't work well. For example: 

1. In case of `scroll` request, the aggregations are not added in the `SearchContext` though it can be part of the `request` object used by `SearchContext`. [See the usage](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/SearchService.java#L1237) of `includeAggregations` flag in the `parseSource` method.
2. For case of timeseries index where sort request is made on timeseries field we also want to use non concurrent path. Details [here](https://github.com/opensearch-project/OpenSearch/issues/9538). To evaluate that we will need the `SortAndFormats` set in `SearchContext` after parsing the source. The condition will be same as used [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L527) in ContextIndexSearcher, which I will add as a follow-up to this change.

Comment (reta) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: > @reta This method will be a single place to evaluate for request based conditions where we want to fallback to non concurrent path for the request

Fair enough, should the at least accept the request it evaluates the conditions? For example: 

```
public void evaluateSearchConstraints(Request request) {
}
```

that would cleanly indicate two things a) request as the context (as you pointed out, the default is not enough) b) not that important, but could potentially evaluate any other restrictions. The concurrent search part should be represented as `SetOnce<Boolean>`

Comment (sohami) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: We don't need `Request` since it is not used. The evaluation is done on the components stored in Context after parsing is complete. Like for `sort` evaluation will use `SortAndFormats` which is stored in context (after parsing) and not in `Request`.

> The concurrent search part should be represented as SetOnce<Boolean>

Makes sense

Comment (reta) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: Thanks @sohami 

> We don't need Request since it is not used. The evaluation is done on the components stored in Context after parsing is complete.

Do we have a deterministic place when this is done? (sorry haven't look yet). Why I am asking, do we have an opportunity to not have calling yet another method (at some place) but deduct from the context?

Comment (reta) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: So to summarize, @sohami  please correct me, by hooking up with:
 - `parseSource()` 
 - `aggregations(...)`  
 - `sort(...)`
 
we could decide (no manual method call) if the search request could use the concurrent search or not (right now we rely on `aggregations(...)` only but it will be extended later on). Is that right?

Comment (sohami) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: No, I was still thinking to keep this method call `evaluateSearchConstraints()` in `parseSource` where it will evaluate in one place based on all request level parameters set in context. It will be more readable to keep this logic in single place instead of multiple methods like `aggregations` and `sort` and also follow the `SetOnce` semantic for the flag `requestShouldUseConcurrentSearch`. I will prefer to keep `sort` and `aggregations` methods simple setters and not have any other logic.

Comment (reta) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: > I will prefer to keep sort and aggregations methods simple setters and not have any other logic

:100: that would be ideal, but would aggregations + sort be set **before** parseSource?

Comment (sohami) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: > but would aggregations + sort be set before parseSource?

No. Since these operators are part of request it gets set in `parseSource`

Comment (reta) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:891]: Ok, so let's keep the logic this way but use `SetOnce`, thanks @sohami

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java [hunk 11]
```diff
@@ -876,11 +877,23 @@ public Profilers getProfilers() {
...
+        if (aggregations() != null && aggregations().factories() != null) {
+            requestShouldUseConcurrentSearch = aggregations().factories().allFactoriesSupportConcurrentSearch();
+        } else {
+            requestShouldUseConcurrentSearch = concurrentSearchSettingsEnabled;
```
Comment (sohami) [inline server/src/main/java/org/opensearch/search/DefaultSearchContext.java:895]: we can set ` requestShouldUseConcurrentSearch` to `true` by default instead of `concurrentSearchSettingsEnabled` as `shouldUseConcurrentSearch` is checking for both.

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java (hunk 10 above)
Comment (jed326) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java:892]: @reta @sohami thanks for the discussion here! I changed this to use `SetOnce` and changed the default case to set `true`.

**Review by sohami — APPROVED** [review]
